File: c++/0791.cpp
// Link: https://leetcode.com/problems/custom-sort-string/description/
// Runtime: 0 ms
// Memory: 7.98 MB

class Solution 
{
public:
    std::string customSortString(const std::string& order, const std::string& s) 
    {
        std::unordered_map<char, int> characterAndFrequency;
        for (const auto& symbol : s)
            characterAndFrequency[symbol]++;
        
        std::string sortedString;
        for (const auto& symbol : order)
            if (characterAndFrequency[symbol] > 0)
            {
                sortedString += std::string(characterAndFrequency[symbol], symbol);
                characterAndFrequency[symbol] = 0;
            }
        
        for (const auto&[key, val] : characterAndFrequency)
            if (val > 0)
                sortedString += std::string(val, key);

        return sortedString;  
    }
};

File: python/0791.py
# Link: https://leetcode.com/problems/custom-sort-string/description/
# Runtime: 37 ms
# Memory: 16.59 MB

class Solution:
    def customSortString(self, order: str, s: str) -> str:
        character_and_frequency = Counter(s)
        sorted_string = ''
        for symbol in order:
            if character_and_frequency[symbol] > 0:
                sorted_string += symbol * character_and_frequency[symbol]
                character_and_frequency[symbol] = 0
        
        for symbol in character_and_frequency:
            sorted_string += symbol * character_and_frequency[symbol]
        
        return sorted_string
